### PR TITLE
Allow creation of pipelines with schema and their visualization

### PR DESCRIPTION
**Pull request by Pavinati**

### CODE REVIEW

File: src/react/PipelineSourcePage.js [hunk 1]
```diff
@@ -95,20 +95,32 @@ export default class PipelineSourcePage extends React.Component {
...
 
         innerHTML = (
           <>
           <Row>
             <Col>
               <h5 className="mt-2 mb-2">Name</h5>
               <p>{name}</p>
-              <h5 className="mt-2 mb-2">Description</h5>
-              <p>{description}</p>
+              { description && description !== '' &&
```
Comment (davidebriani) [inline src/react/PipelineSourcePage.js:106]: `description && description !== '' && ...` is the same as `description && ...` so you could just use the latter. The same goes for `schema` below

File: src/react/PipelineSourcePage.js [hunk 2]
```diff
@@ -95,20 +95,32 @@ export default class PipelineSourcePage extends React.Component {
...
+              }
               <h5 className="mt-2 mb-2">Source</h5>
               <SyntaxHighlighter language="text" showLineNumbers="true">
                 {source}
               </SyntaxHighlighter>
+              { schema &&
+                <>
+                  <h5 className="mt-2 mb-2">Schema</h5>
+                  <SyntaxHighlighter language="text" showLineNumbers="true">
+                    {JSON.stringify(schema, null, '  ')}
```
Comment (davidebriani) [inline src/react/PipelineSourcePage.js:120]: I think `JSON.stringify(schema, null, 2)` might be clearer

File: src/react/PipelineSourcePage.js [hunk 3]
```diff
@@ -95,20 +95,32 @@ export default class PipelineSourcePage extends React.Component {
...
+                </>
+              }
               <h5 className="mt-2 mb-2">Source</h5>
               <SyntaxHighlighter language="text" showLineNumbers="true">
                 {source}
               </SyntaxHighlighter>
+              { schema &&
+                <>
+                  <h5 className="mt-2 mb-2">Schema</h5>
+                  <SyntaxHighlighter language="text" showLineNumbers="true">
```
Comment (davidebriani) [inline src/react/PipelineSourcePage.js:119]: Why not setting the language as `json` instead of `text`?

File: src/react/NewPipelinePage.js [hunk 4]
```diff
@@ -16,145 +16,159 @@
...
+            as="textarea"
+            value={pipelineDescription}
+            onChange={(e) => setPipelineDescription(e.target.value)}
+          />
+        </Form.Group>
+      </Form>
+      <Button
+        variant="primary"
+        onClick={createPipeline}
+        disabled={pipelineName === '' || pipelineName === 'new' || pipelineSource === ''}
```
Comment (Pavinati) [inline src/react/NewPipelinePage.js:159]: add `isCreatingPipeline` to disabled condition

Comment (davidebriani) [reply src/react/NewPipelinePage.js:159]: You could define a `isValidForm` to keep this line cleaner

File: src/react/NewPipelinePage.js [hunk 5]
```diff
@@ -16,145 +16,159 @@
...
+let alertId = 0;
+const ajv = new Ajv({schemaId: 'id'});
+ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-04.json'));
+
+export default ({ astarte, history, blockId }) => {
+  const [alerts, setAlerts] = useState(new Map());
+  const [pipelineName, setPipelineName] = useState("");
+  const [pipelineSource, setPipelineSource] = useState("");
+  const [pipelineSchema, setPipelineSchema] = useState("");
+  const [pipelineDescription, setPipelineDescription] = useState("");
```
Comment (davidebriani) [inline src/react/NewPipelinePage.js:40]: Maybe we could decide, for all form pages, how to keep the state. For instance, just defining a single variable could help clean up the code:

```js
const [pipeline, setPipeline] = useState({
    name: "",
    source: "",
    schema: "",
    description: ""
  });
```

```jsx
<Form.Control
    value={pipeline.name}
    onChange={(e) => setPipeline({ ...pipeline, name: e.target.value})}
>
```

Comment (Pavinati) [reply src/react/NewPipelinePage.js:40]: If you do that, you lose control on the single component changes. In this specific example, you will rebuild the schema object every time the `pipeline` changes, even if only the name/description changed.

Comment (Pavinati) [reply src/react/NewPipelinePage.js:40]: Disregard previous comment, on further inspection, the dependency array can contain any value. Even a subset of the used object.

Comment (davidebriani) [reply src/react/NewPipelinePage.js]: Yes, exactly. If we actually had to extract a Form component, it is common to see in libraries handling forms that a single object is passed to the form to inject/hydrate form's input values. So keeping an entity's state together is not a new pattern. I'll leave this as point open for future discussion

**Pull request merged**